Applications/Custom/momentum.cpp

```diff
@@ -6,55 +6,48 @@
  * @date   1 June 2021
  * @see    https://github.com/nnstreamer/nntrainer
  * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
  * @bug    No known bugs except for NYI items
  * @brief  This is the Momentum optimizer.
  */
-
-#include <cmath>
-#include <fstream>
-
 #include <momentum.h>
 #include <nntrainer_error.h>
 #include <nntrainer_log.h>
+#include <node_exporter.h>
+#include <tensor.h>
 
-namespace nntrainer {
-
-const std::string Momentum::type = "momentum";
+namespace custom {
 
-void Momentum::addOptimizerVariable(std::vector<Weight> &weight_list) {
-  for (auto &w : weight_list) {
-    w.clearOptimizerVariables();
+Momentum::Momentum() { setProperty({"learning_rate=0.001", "momentum=0.9f"}); }
 
-    if (!w.getTrainable())
-      continue;
-
-    w.addOptimizerVariable(w.getDim()); /** Add accumulated_momentum */
-  }
-}
+void Momentum::applyGradient(nntrainer::RunOptimizerContext &context) {
+  nntrainer::Tensor &x_grad = context.getGradient();
+  nntrainer::Tensor &accumulated = context.getOptimizerVariable(0);
 
-void Momentum::applyGradient(Weight &weight, double updated_lr, int iteration) {
+  float momentum = std::get<PropMomentum>(momentum_props);
 
-  Tensor &x_grad = weight.getGradientRef();
-  Tensor &accumulated = weight.getOptimizerVariableRef(0);
-
-  accumulated.multiply_i(momentum);
-  accumulated.add_i(x_grad);
+  accumulated.add_i(x_grad, momentum);
 
   x_grad.fill(accumulated);
-  weight.applyGradient(updated_lr);
+  context.applyGradient(getLearningRate(context.getIteration()));
+}
+
+double Momentum::getLearningRate(size_t iteration) const {
+  return nntrainer::OptimizerImpl::getLearningRate(iteration);
 }
 
-void Momentum::setProperty(const std::string &key, const std::string &value) {
-  int status = ML_ERROR_NONE;
-  if (key == "momentum") {
-    status = setDouble(momentum, value);
-  } else {
-    OptimizerImpl::setProperty(key, value);
-  }
+std::vector<ml::train::TensorDim>
+Momentum::getOptimizerVariableDim(const ml::train::TensorDim &dim) {
+  /// momentum optimizer uses *accumulated
+  return {dim};
```

Comment:
This variable is allocated for each weight with ITERATION_LIFESPAN.
How about give a lifespan option (differentiate between local and iteration memory)?

Recommend to do in a separate PR.

Reply:
Got it, just confirming, isn't this max lifespan == iteration lifespan (as of now?)

Reply:
no , iteration lifespan != max lifespan.